chanzuckerberg/idseq-web
# Precache report info on success of pipeline run

# Description

This fixes the bad perf of the sample report page by caching the http response as a file on disk by using the Rails standard action caching. The browser HTTP client-side caching behavior is preserved in prod, and enabled in dev.

My tests of a large sample show 
1) report_info time decreased from 8s to 1s by server side caching
2) report_info time decreased to less than 1s by client side caching

I refactored some of the `report_info` and child methods to make the param dependencies more clear.

I still have some questions about the correct defaults for params like `background_id` which will affect the cache hit rate. Note we can always precache with more param values. 

NOTE: the choice of cache keys means that the server cache of a report will be invalidated on 
* new pipeline run
* any update of sample object
* new code push

# Test and Reproduce

Test caching
1. Open a report such as http://localhost:3000/samples/12303
2. See SQL queries in rails log 
3. Reload the report 
4. See cache line in rails log such as 

```
ActiveSupport::Cache::FileStore : Cache read: /tmp/rails_cache_store/BD7/1E3/views%2Flocalhost%3A3000%2Fsamples%2F10399%2Freport_info.json%3Fbackground_id%3D29%26git_version%3D%26pipeline_version%3D3.3%26report_ts%3D1548725583%26sort_by%3Dnt_aggregatescore ({:expires_in=>365 days})
```

5. Verify that the decoded URL is correctly unique in param values
```/tmp/rails_cache_store/BD7/1E3/views/localhost:3000/samples/10399/report_info.json?background_id=29&git_version=&pipeline_version=3.3&report_ts=1548725583&sort_by=nt_aggregatescore```

6. Verify cache on further reloads of report page

7. Verify cache miss on changing background


Test precaching:

1.  In `rails console`, load a pipeline run
2. Precache  method and see the following:

```
irb(main):002:0> PipelineRun.find(18660).precache_report_info!
...
[2019-05-07T19:28:38] DEBUG Rails : Precaching URL http://localhost:3000/samples/12295/report_info?background_id=93&format=json&git_version=&pipeline_version=3.3&report_ts=1549325153&scoring_model=aggregate_score&sort_by=nt_aggregatescore
=> #<Tempfile:/tmp/open-uri20190507-144-u0xcg3>
```
3. Run it again and see a cache hit in the rails logs


Overall perf testing:

I'll review the stats before and after launch to prod in the airbrake dashboard.
